### ISSUE TITLE
Question concerning lang attribute

### ISSUE THREAD

**Issue by ReLater** [mnater/Hyphenopoly#29]
Just a question if you have the time to answer.
Example: If a page starts with
```
<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="de-de" lang="de-de" >
```
(no other lang attributes) does Hyphenopoly understand that the pattern `de.hpb` is relevant for the whole page. Or do I have to add a `lang="de"` explicitely (to BODY tag or so)? Or duplicate the pattern file to `de-de.hpb` ?

The same with example `en-au` or similiar. Is there an automatic fallback to `en-gb` or other `en-xy` pattern? Or to `en.hpb` if there would be one?

Thank you!

**Comment by mnater** [comment]
Hi

As of the actual version 2.3.0 Hyphenopoly needs an exact match of the lang attribute with the pattern file name. So the answer is no.

Both approaches should work (eihter adding an other `lang="de"` or duplicating and renaming the pattern file) although I'd recommand the latter to keep the html clean.

Best regards,
Mathias

**Comment by mnater** [comment]
I'm leaving this open, because after all this is a bug.

To do:
- [x] investigate how current UA's cope with extended lang tags;
- [x] cope with `lang` attributes that don't exactly match the existing pattern names;
- [x] elaborate a mapping mechanism between the two;
- [x] document the behaviour.

**Comment by ReLater** [comment]
Thanks a lot for your efforts and explanation! A configurable mapping would be great.

I'm testing with my custom Joomla plugin (based upon Hyphenator) where users can select several languages from a list box based upon the files in `/patterns/` and can add the "long words". Would be nice to give them the opportunity to define fallbacks. Something like
`en-au` => `en-gb`

If Hyphenopoly would understand these mappings directly, really welcome, but it would be also possible for me to copy a pattern automatically via plugin PHP if a user needs one.

**Comment by mnater** [comment]
Hyphenator makes a xmlHttpRequest (HEAD) to check if the pattern is available and then falls back to a more general pattern, if the check fails.

I investigated a little how UA's cope with extended lang tags:

__tldr: it's a bag of hurt!__

FF for example uses `en-us` for all sorts of english (Mac and Win).
Safari and Chrome (Mac) use the patterns provided by OS X, where patterns for `en-us` and `en-gb` are available. For `en-au` Safari and Chrome both fall back to `en-us`.

I didn't investigate the behaviour of IE and Edge.

https://tools.ietf.org/html/rfc4647 defines the BCP for _Matching of Language Tags_:

````
Example of a Lookup Fallback Pattern

   Range to match: zh-Hant-CN-x-private1-private2
   1. zh-Hant-CN-x-private1-private2
   2. zh-Hant-CN-x-private1
   3. zh-Hant-CN
   4. zh-Hant
   5. zh
   6. (default)
````

One option would be to have a list of supported languages and defined fallbacks in `Hyphenopoly_Loader`. An other option would be to make a HEAD-request (just getting the http HEAD, not the content) to see if patterns are available.

In Hyphenopoly I'd like to keep the code and the numbers of requests as small as possible (for performance), so both options are a no go.

So the only option is to extend the API of Hyphenopoly to define fallbacks (as you requested):

````javascript
cons Hyphenopoly = {
            require: {
                "en-au": "supercalifragilisticexpialidocious"
            },
            fallbacks: {
                "en-au": "en-gb"
            }
        };
````

(I won't have time to work on this for a the next three weeks…)

**Comment by mnater** [comment]
Implemented `fallbacks` (961c15bbef3a634964e2adc222093b69bf6ea53d)
Still needs documentation!

**Comment by mnater** [comment]
Doc: https://github.com/mnater/Hyphenopoly/wiki/Global-Hyphenopoly-Object#fallbacks
